src/Tests/InstrEngineTests/InstrEngineTests/EmbeddedResourceUtils.cs

```diff
@@ -0,0 +1,160 @@
+                {
+                    stream.CopyTo(writeStream);
+                }
+            }
+        }
+
+        public static string ReadEmbeddedResourceFile(string file, bool required = true)
+        {
+            Stream stream = GetEmbeddedResource(file, required);
+            if (stream == null)
+            {
+                return null;
+            }
+
+            using (StreamReader reader = new StreamReader(stream))
+            {
+                return reader.ReadToEnd();
+            }
+        }
+
+        /// <summary>
+        /// Looks for the given embedded resource, and extracts it to the test assets folder.
+        /// If the file was previously extracted, it is not overwritten.
+        /// </summary>
+        /// <param name="resourceName">The name of the resource.</param>
+        /// <param name="required">True if an error should occur when the resource is not found.</param>
+        /// <returns>An IEmbenddedResourceFile that has info about the extracted file.</returns>
+        public static IEmbeddedResourceFile GetTestResourceFile(string resourceName, bool required = true)
+        {
+            // Don't allow parallel tests to overwrite extracted files.
+            lock (ExtractionLock)
+            {
+                Stream stream = GetEmbeddedResource(resourceName, required);
+                if (stream == null)
+                {
+                    return null;
+                }
+                using (stream)
+                {
+                    DirectoryInfo resourceDirectory = new DirectoryInfo(PathUtils.GetResourcesFolder());
+                    resourceDirectory.Create();
+
+                    string tempFileName = Path.Combine(PathUtils.GetResourcesFolder(), resourceName);
+                    FileInfo fileInfo = new FileInfo(tempFileName);
+                    var embeddedResource = new EmbeddedResourceFile(fileInfo, resourceName);
+                    try
+                    {
+                        if (!fileInfo.Exists)
+                        {
+                            embeddedResource.CopyFromStream(stream);
+                        }
+
+                        return embeddedResource;
+                    }
+                    catch (Exception e)
+                    {
+                        Assert.Fail(e.Message);
+                    }
+                }
+
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Deletes all test resource files that have been previously extracted.
+        /// </summary>
+        public static void CleanTestResourceFiles()
+        {
+            lock (ExtractionLock)
+            {
+                DirectoryInfo resourceDirectory = new DirectoryInfo(PathUtils.GetResourcesFolder());
+                if (resourceDirectory.Exists)
+                {
+                    foreach (FileInfo f in resourceDirectory.EnumerateFiles())
+                    {
+                        f.Delete();
```

Comment:
Something to consider - do we want/need to handle potential `IOExceptions` here, where a deletion might fail?

Reply:
I think we can leave it alone. The tests shouldn't pass if the delete fails. There is something wrong with the infrastructure in that case.